File: paper/idiomatic/cpg.bio.jl
using BioSequences

m, M = 99999, 0
function cpg(s)
   i, cnt = 1, 0
   mm, MM = 99999, 0
   while i <= length(s)
      cc, l = 0, 0
      while i <= length(s) 
         cc |= if s[i] == DNA_C 1 elseif s[i] == DNA_G 2 else break end
         i += 1
         l += 1
      end
      if cc == 3
         cnt += 1
         mm = min(mm,l)
         MM = max(MM,l)
      end
      i += 1
   end
   global m = min(m,mm)
   global M = max(M,MM)
   cnt
end

function main(arg)
   f = open(arg)
   cnt = 0
   for l in eachline(f)
      cnt += cpg(DNASequence(l))
   end
   println("$cnt $m $M")
end

main(ARGS[1])

File: paper/idiomatic/cpg.bio.py
import sys
import Bio.Seq

def cpg(s):
   for c in s:
      i = 1 if c == 'C' else (2 if c == 'G' else 0)
      if i == 0: break
      yield i

m, M = 99999, 0
def cpg_count(s):
   mm, MM = 99999, 0
   i = 0
   count = 0
   while i < len(s):
      cc, l = 0, 0
      for j in cpg(s[i:]):
         cc |= j
         l += 1
      if cc == 3:
         count += 1
         mm = min(mm, l)
         MM = max(MM, l)
      i += l + 1
   global m, M
   m = min(mm, m)
   M = max(MM, M)
   return count

f = open(sys.argv[1], 'r')
cnt = 0
for l in f:
   cnt += cpg_count(Bio.Seq.Seq(l.strip()))
print cnt, m, M
